### PR TITLE
PKSC7 firmware revision + callback / get SID

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**

### CODE REVIEW

File: wolfcrypt/test/test.c [hunk 1]
```diff
@@ -21286,6 +21398,240 @@ int pkcs7authenveloped_test(void)
...
+        word32 encryptKeySz, byte keyHint, byte* cert, word32 certSz,
+        byte* key, word32 keySz)
+{
+    int ret;
+    PKCS7* pkcs7;
+
+    /* KEY ID
+     * fwDecryptKeyID OID 1.2.840.113549.1.9.16.2.37
+     */
+    unsigned char keyOID[] = {
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: Consider const for keyOID, keyID and data variables.

File: wolfcrypt/test/test.c [hunk 2]
```diff
@@ -21286,6 +21398,240 @@ int pkcs7authenveloped_test(void)
...
+        0x06, 0x0B,
+        0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D,
+        0x01, 0x09, 0x10, 0x02, 0x25
+    };
+    byte keyID[] = { 0x04, 0x01, 0x00 };
+    byte data[] = "Test of wolfSSL PKCS7 decrypt callback";
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: You'll get a cast warning with some compilers on char to unsigned char here.

File: wolfcrypt/test/test.c [hunk 3]
```diff
@@ -20164,6 +20165,117 @@ static int myOriDecryptCb(PKCS7* pkcs7, byte* oriType, word32 oriTypeSz,
...
+    int keyId = -1, ret, keySz;
+    word32 keyIdSz = 8;
+    byte*  key;
+    byte   keyIdRaw[8];
+    Aes    aes;
+
+    /* looking for KEY ID
+     * fwDecryptKeyID OID "1.2.840.113549.1.9.16.2.37
+     */
+    unsigned char OID[] = {
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: Consider const on OID, defKey, altKey.

File: wolfcrypt/test/test.c [hunk 4]
```diff
@@ -20164,6 +20165,117 @@ static int myOriDecryptCb(PKCS7* pkcs7, byte* oriType, word32 oriTypeSz,
...
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08
+    };
+
+    byte altKey[] = {
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,
+        0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08
+    };
+
+    /* test user context passed in */
+    if (*(int*)usrCtx != 1) {
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: Consider NULL check here.

File: wolfcrypt/test/test.c [hunk 5]
```diff
@@ -20164,6 +20165,117 @@ static int myOriDecryptCb(PKCS7* pkcs7, byte* oriType, word32 oriTypeSz,
...
+    /* if needing to find keyIdSz can call with NULL */
+    ret = wc_PKCS7_GetAttributeValue(pkcs7, OID, sizeof(OID), NULL,
+            &keyIdSz);
+    if (ret != LENGTH_ONLY_E) {
+        printf("Unexpected error %d when getting keyIdSz\n", ret);
+        printf("Possibly no KEY ID attribute set\n");
+        return -1;
+    }
+    else {
+        memset(keyIdRaw, 0, sizeof(keyIdRaw));
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: Let's use XMEMSET for wolfCrypt test portability.

File: wolfcrypt/test/test.c [hunk 6]
```diff
@@ -21286,6 +21398,240 @@ int pkcs7authenveloped_test(void)
...
+    int ret = 0;
+    int usrCtx = 1; /* test value to pass as user context to callback */
+    PKCS7 pkcs7;
+    byte*  sid;
+    word32 sidSz;
+
+    byte decoded[FOURK_BUF/2];
+    int  decodedSz = FOURK_BUF/2;
+
+    byte expectedSid[] = {
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: Consider const for expectedSid.

File: wolfcrypt/test/test.c [hunk 7]
```diff
@@ -21286,6 +21398,240 @@ int pkcs7authenveloped_test(void)
...
+
+
+int pkcs7callback_test(byte* cert, word32 certSz, byte* key, word32 keySz)
+{
+
+    int ret = 0;
+    byte derBuf[FOURK_BUF/2];
+    word32 derSz = FOURK_BUF/2;
+
+    byte defKey[] = {
```
Comment (dgarske) [inline wolfcrypt/test/test.c]: Consider const for defKey and altKey.

**Comment by JacobBarthelmeh** [comment]
Thanks for the initial review, I'll go through and address the items brought up.

**Pull request merged**